pg_lakehouse/src/datafusion/session.rs

```diff
@@ -1,59 +1,172 @@
 use datafusion::prelude::{SessionConfig, SessionContext};
 use once_cell::sync::Lazy;
 use pgrx::*;
-use std::collections::{
-    hash_map::Entry::{Occupied, Vacant},
-    HashMap,
-};
+use std::collections::{hash_map::Entry::Vacant, HashMap};
 use std::ffi::CStr;
-use std::future::Future;
-use std::pin::Pin;
 use std::sync::Arc;
+use thiserror::Error;
 
-use super::context::ContextError;
+use super::catalog::*;
+use super::schema::LakehouseSchemaProvider;
 
 const SESSION_ID: &str = "lakehouse_session_context";
 
-static SESSION_CACHE: Lazy<Arc<Mutex<HashMap<String, SessionContext>>>> =
-    Lazy::new(|| Arc::new(Mutex::new(HashMap::new())));
+type SessionCache = Lazy<Arc<RwLock<HashMap<String, Arc<SessionContext>>>>>;
+static SESSION_CACHE: SessionCache = Lazy::new(|| Arc::new(RwLock::new(HashMap::new())));
 
 pub struct Session;
 
 impl Session {
-    pub fn with_session_context<F, R>(f: F) -> Result<R, ContextError>
-    where
-        F: for<'a> FnOnce(
-            &'a SessionContext,
-        ) -> Pin<Box<dyn Future<Output = Result<R, ContextError>> + 'a>>,
-    {
-        let mut lock = task::block_on(SESSION_CACHE.lock());
-
-        let context = match lock.entry(SESSION_ID.to_string()) {
-            Occupied(entry) => entry.into_mut(),
-            Vacant(entry) => {
-                // Set current timezone
-                let mut session_config = SessionConfig::from_env()?.with_information_schema(true);
-                let session_timezone = unsafe {
-                    CStr::from_ptr(pg_sys::pg_get_timezone_name(pg_sys::session_timezone))
-                        .to_str()
-                        .unwrap_or_else(|err| panic!("{:?}", err))
-                };
-                session_config.options_mut().execution.time_zone =
-                    Some(session_timezone.to_string());
-
-                // Create a new context
-                let rn_config = RuntimeConfig::new();
-                let runtime_env = RuntimeEnv::new(rn_config)?;
-                let context =
-                    SessionContext::new_with_config_rt(session_config, Arc::new(runtime_env));
-
-                entry.insert(context)
-            }
+    pub fn session_context() -> Result<Arc<SessionContext>, SessionError> {
+        let mut write_lock = task::block_on(SESSION_CACHE.write());
+
+        if let Vacant(entry) = write_lock.entry(SESSION_ID.to_string()) {
+            entry.insert(Self::init()?);
+        }
+
+        drop(write_lock);
+
+        let read_lock = task::block_on(SESSION_CACHE.read());
+        let context = read_lock
+            .get(SESSION_ID)
+            .ok_or(SessionError::ContextNotFound)?;
+
+        Ok(context.clone())
+    }
+
+    pub fn catalog() -> Result<LakehouseCatalog, SessionError> {
+        let mut write_lock = task::block_on(SESSION_CACHE.write());
+
+        if let Vacant(entry) = write_lock.entry(SESSION_ID.to_string()) {
+            entry.insert(Self::init()?);
+        }
+
+        drop(write_lock);
```

Comment:
I think it's better practice to just create a block when you want something to be dropped:

```rust
{
        let mut write_lock = task::block_on(SESSION_CACHE.write());

        if let Vacant(entry) = write_lock.entry(SESSION_ID.to_string()) {
            entry.insert(Self::init()?);
        }
}
```

Stack variables get dropped at the end of any block.

Reply:
I like that!